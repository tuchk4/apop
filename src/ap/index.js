import crop from './actions/splice';
import filter from './actions/filter';
import join from './actions/join';
import map from './actions/map';
import remove from './actions/remove';
import sort from './actions/sort';
import swap from './actions/swap';

import { apply, recursiveApply } from './utils/apply';

const sourceActions = { crop, filter, join, map, remove, sort, swap };

const ap = Object.assign(apply, sourceActions);
const recursive = Object.assign(recursiveApply, sourceActions);

export default Object.assign(ap, { recursive });
