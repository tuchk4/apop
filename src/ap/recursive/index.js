import crop from '../actions/splice';
import filter from '../actions/filter';
import join from '../actions/join';
import map from '../actions/map';
import remove from '../actions/remove';
import sort from '../actions/sort';
import swap from '../actions/swap';

import { recursiveApply } from '../utils/apply';

const sourceActions = { crop, filter, join, map, remove, sort, swap };
const recursiveActions = {};

Object.keys(sourceActions).forEach(action => {
  recursiveActions[action] = (...args) =>
    recursiveApply(sourceActions[action](...args));
});

const recursive = Object.assign(
  (...actions) => recursiveApply(...actions),
  recursiveActions
);

export default recursive;
