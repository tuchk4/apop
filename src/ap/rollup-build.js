// import { rollup } from 'rollup';
//
// import resolve from 'rollup-plugin-node-resolve';
// import uglify from 'rollup-plugin-uglify';
// import commonjs from 'rollup-plugin-commonjs';
// import babel from 'rollup-plugin-babel';
// import { minify } from 'uglify-es';
// import babelrc from 'babelrc-rollup';
//
// const babelConfig = {
//   presets: [['es2015', { loose: true, modules: false }, 'stage-0']],
//   plugins: ['external-helpers', 'transform-object-rest-spread'],
// };
//
// var rollUpConfig = {
//   format: 'umd',
//   sourceMap: 'inline',
//   plugins: [
//     resolve({
//       jsnext: true,
//       main: true,
//       browser: true,
//     }),
//     commonjs(),
//     babel(
//       babelrc({
//         addExternalHelpersPlugin: false,
//         config: babelConfig,
//         exclude: 'node_modules/**',
//       })
//     ),
//   ],
// };
//
// async function make(config) {
//   let bundle = await rollup({
//     entry: `./src/${config.moduleName}.js`,
//     plugins: config.plugins,
//   });
//
//   await bundle.write({
//     format: config.format,
//     moduleName: config.moduleName,
//     dest: `./${config.moduleName}.js`,
//     sourceMap: config.sourceMap,
//   });
//   await bundle.write({
//     format: config.format,
//     moduleName: config.moduleName,
//     dest: `./${config.moduleName}.min.js`,
//     sourceMap: config.sourceMap,
//     plugins: [...rollUpConfig.plugins, uglify({}, minify)],
//   });
//   console.log(`make ./${config.moduleName}.js and ./${config.moduleName}.min.js`);
// }
//
// (async function() {
//
//   const builds = ['rmkArray','recursive/rmkArray'];
//
//   builds.forEach(async moduleName => await make({...rollUpConfig, moduleName}));
//
// })();
