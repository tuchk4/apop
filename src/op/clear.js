import action from './actions/clear';
import { apply } from './utils/apply';

export default config => apply(action(config));
