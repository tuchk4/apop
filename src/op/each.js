import action from './actions/each';
import { apply } from './utils/apply';

export default config => apply(action(config));
