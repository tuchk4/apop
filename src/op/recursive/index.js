import actions from '../actions';

import { recursiveApply } from '../utils/apply';

const recursive = Object.assign((...args) => recursiveApply(...args), actions);

export default recursive;
