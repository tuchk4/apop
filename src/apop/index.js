import op from '../op/index';
import ap from '../ap/index';
import opRecursive from '../op/recursive';
import apRecursive from '../ap/recursive';
import apply from './utils/apply';

import clear from '../op/actions/clear';
import toSnakeCase from '../op/actions/toSnakeCase';
import toCamelCase from '../op/actions/toCamelCase';
import update from '../op/actions/update';
import rename from '../op/actions/rename';
import each from '../op/actions/each';

const sourceActions = { clear, toSnakeCase, toCamelCase, update, rename, each };
const singleActions = {};

Object.keys(sourceActions).forEach(action => {
  singleActions[action] = (...args) => apply(sourceActions[action](...args));
});

const recursive = {
  op: opRecursive,
  ap: apRecursive,
};

export default Object.assign(
  (...actions) => apply(...actions),
  singleActions,
  op,
  ap,
  recursive
);
