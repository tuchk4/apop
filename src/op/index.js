import clear from './actions/clear';
import toSnakeCase from './actions/toSnakeCase';
import toCamelCase from './actions/toCamelCase';
import update from './actions/update';
import rename from './actions/rename';
import each from './actions/each';

import { apply, recursiveApply } from './utils/apply';

const sourceActions = { clear, toSnakeCase, toCamelCase, update, rename, each };
const hackActions = {};

Object.keys(sourceActions).forEach(key => {
  hackActions[key] = (...args) =>
    Object.assign(sourceActions[key](...args), { shortcut: true, key });
});

const op = Object.assign((...args) => apply(...args), hackActions);
const recursive = Object.assign(
  (...args) => recursiveApply(...args),
  hackActions
);

export default Object.assign(op, { recursive });
