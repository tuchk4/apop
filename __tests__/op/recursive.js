import ap from '../../src/build/ap';
import op from '../../src/build/op';
import apop from '../../src/build/apop';

const sample = Object.seal({
  foo_bar: 0,
  remove_baz: null,
  array_key: [1, 2, 3],
  array_objects: [{ a_b: 'a', b_c: 'b' }],
});

const data = Object.seal({
  asObject: Object.assign({ deep_obj: sample }, sample),
  asArray: [
    {
      children: [],
      name: '1',
      url: '/1',
    },
    {
      children: [],
      name: '2',
      url: '/v2/2',
    },
    {
      children: [
        {
          children: [],
          name: '3',
          url: '/v2/3',
        },
        {
          children: [
            {
              children: [],
              name: '4',
              url: '/4',
            },
            {
              children: [],
              name: '5',
              url: '/v2/5',
            },
            {
              children: [],
              name: '6',
              url: '/v2/6',
            },
          ],
          name: '7',
          url: '/7',
        },
      ],
      name: '8',
      url: '/8',
    },
  ],
});

const VERSIONS = Object.seal({
  V1: 'v1',
  V2: 'v2',
});

const versionLinkFn = (url, isLink) => {
  let version;
  if (url && isLink) {
    version = url.startsWith(`/${VERSIONS.V2}`) ? VERSIONS.V2 : VERSIONS.V1;
  }

  return version;
};

const normalizeUrl = (url, version) =>
  version === VERSIONS.V2 ? url.replace(`/${VERSIONS.V2}`, '') : url;

test('Object recursive', () => {
  const formula = op.recursive(
    op.toCamelCase(),
    op.update({
      arrayKeyStr: localState =>
        Array.isArray(localState.arrayKey)
          ? localState.arrayKey.join(',')
          : null,
    }),
    op.rename({
      arrayKeyStr: 'renamedStr',
    }),
    op.clear()
  );

  const result = formula(data.asObject);

  expect(result).toMatchSnapshot();
});
let sum = '';
test('Array recursive', () => {
  const transformMainMenu = op.recursive(
    op.clear(),
    op.toCamelCase(),
    op.rename({
      children: 'nodes',
    }),
    op.update({
      isCategory: localState =>
        Object.prototype.hasOwnProperty.call(localState, 'nodes'),
      isLink: localState =>
        !Object.prototype.hasOwnProperty.call(localState, 'nodes'),
    }),
    op.update({
      version: localState => versionLinkFn(localState.url, localState.isLink),
    }),
    op.update({
      url: localState => normalizeUrl(localState.url, localState.version),
      name: localState => {
        sum += `->${localState.name}`;
        return `!${localState.name}`;
      },
      isOpen: localState => {
        let isOpen = null;
        if (localState.isCategory) {
          isOpen = false;
        }
        return isOpen;
      },
    }),
    op.clear()
  );
  const result = transformMainMenu(data.asArray);

  expect(result).toMatchSnapshot();
  expect(sum).toEqual('->1->2->8->3->7->4->5->6');
});

it('recursive shortcut rename', () => {
  const result = op.recursive(op.rename({ name: 'test' }))(data.asArray);
  expect(result).toMatchSnapshot();
});

it('recursive shortcut toCamelCase', () => {
  const result = op.recursive(op.toCamelCase())(sample);
  expect(result).toMatchSnapshot();
});
