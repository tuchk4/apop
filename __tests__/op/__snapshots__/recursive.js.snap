// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array recursive 1`] = `
Array [
  Object {
    "isCategory": false,
    "isLink": true,
    "name": "!1",
    "url": "/1",
    "version": "v1",
  },
  Object {
    "isCategory": false,
    "isLink": true,
    "name": "!2",
    "url": "/2",
    "version": "v2",
  },
  Object {
    "isCategory": true,
    "isLink": false,
    "isOpen": false,
    "name": "!8",
    "nodes": Array [
      Object {
        "isCategory": false,
        "isLink": true,
        "name": "!3",
        "url": "/3",
        "version": "v2",
      },
      Object {
        "isCategory": true,
        "isLink": false,
        "isOpen": false,
        "name": "!7",
        "nodes": Array [
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "!4",
            "url": "/4",
            "version": "v1",
          },
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "!5",
            "url": "/5",
            "version": "v2",
          },
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "!6",
            "url": "/6",
            "version": "v2",
          },
        ],
        "url": "/7",
      },
    ],
    "url": "/8",
  },
]
`;

exports[`Object recursive 1`] = `
Object {
  "arrayKey": Array [
    1,
    2,
    3,
  ],
  "arrayObjects": Array [
    Object {
      "aB": "a",
      "bC": "b",
    },
  ],
  "deepObj": Object {
    "arrayKey": Array [
      1,
      2,
      3,
    ],
    "arrayObjects": Array [
      Object {
        "aB": "a",
        "bC": "b",
      },
    ],
    "fooBar": 0,
    "renamedStr": "1,2,3",
  },
  "fooBar": 0,
  "renamedStr": "1,2,3",
}
`;

exports[`recursive shortcut rename 1`] = `
Array [
  Object {
    "children": Array [],
    "test": "1",
    "url": "/1",
  },
  Object {
    "children": Array [],
    "test": "2",
    "url": "/v2/2",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "test": "3",
        "url": "/v2/3",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "test": "4",
            "url": "/4",
          },
          Object {
            "children": Array [],
            "test": "5",
            "url": "/v2/5",
          },
          Object {
            "children": Array [],
            "test": "6",
            "url": "/v2/6",
          },
        ],
        "test": "7",
        "url": "/7",
      },
    ],
    "test": "8",
    "url": "/8",
  },
]
`;

exports[`recursive shortcut toCamelCase 1`] = `
Object {
  "arrayKey": Array [
    1,
    2,
    3,
  ],
  "arrayObjects": Array [
    Object {
      "aB": "a",
      "bC": "b",
    },
  ],
  "fooBar": 0,
  "removeBaz": null,
}
`;
